### PR TITLE
Config spec v0.0.2 proposal

### PR DESCRIPTION
**Pull request by hktouw**
## Why

To support more sophisticated workflows that need multiple stages, stored queries, and state management we should update the config spec to support these features.

## How

Please read the PR for a proposal for how to support these features. 

## Feedback requested

- General feedback
- Does this approach make sense? Is there a better way to accomplish this?
- What are considerations that are not being addressed here? Does this paint us into a corner in undesirable ways?

### CODE REVIEW

File: config-spec.md [hunk 1]
```diff
@@ -65,3 +67,71 @@ workflow:
...
+
+```yaml
+// Worfklow name: Get all issues for a given user
+// This is a multi-stage workflow
+// 1. Show input form to add account_id
+// 2. User inputs account_id
+// 3. Shows list of issues belonging to that account_id
+workflow:
+  version: 1
+  signal:
```
Comment (kmcq) [inline config-spec.md:107]:
```suggestion
  signals:
```

**Review by kmcq — COMMENTED** [review]
Really interesting stuff! Question about the versioning: as far as I can tell this includes no breaking changes to the existing spec. Do you see this causing us to version the engine? Right now it doesn't really matter, but I'd love to limit versioning the engine as much as possible as it might be annoying

**Comment by hktouw** [comment]
> Question about the versioning: as far as I can tell this includes no breaking changes to the existing spec. Do you see this causing us to version the engine? Right now it doesn't really matter, but I'd love to limit versioning the engine as much as possible as it might be annoying

I don't think anything in here should be a breaking change! And fwiw, I am of the opinion that it is probably fine to do breaking changes until we get to a V1 for the reasons you mentioned.

File: config-spec.md [hunk 2]
```diff
@@ -65,3 +67,71 @@ workflow:
...
+            - :account_id
+    - frame:
+      data:
+        query: >\n
+          SELECT id AS issue_id
+          FROM issues
+          WHERE author_id = :account_id
+        mapColumns:
+          issue_id: issue_ids # TableOfIssues takes issue_ids
+        type: storedQuery
```
Comment (jonmagic) [inline config-spec.md:125]: Not sure why but feeling like this should be capitalized.

```suggestion
        type: StoredQuery
```

File: config-spec.md [hunk 3]
```diff
@@ -65,3 +67,71 @@ workflow:
...
+# Worfklow name: Get all issues for a given user
+# This is a multi-stage workflow
+# 1. Show input form to add account_id
+# 2. User inputs account_id
+# 3. Shows list of issues belonging to that account_id
+workflow:
+  version: 1
+  signals:
+    - account_id: number
+  frameRender: sequence
```
Comment (jonmagic) [inline config-spec.md:109]: Can the value of this be a "component", i.e. a pascal cased function?

```suggestion
  renderFrames: InSequence
```

File: config-spec.md [hunk 4]
```diff
@@ -65,3 +67,71 @@ workflow:
...
+        props:
+          fields:
+            - :account_id
+    - frame:
+      data:
+        query: >\n
+          SELECT id AS issue_id
+          FROM issues
+          WHERE author_id = :account_id
+        mapColumns:
```
Comment (jonmagic) [inline config-spec.md:123]: We talked about changing this to be a `transforms` key that can take in a list of map reduce functions I think?

**Review by jonmagic — APPROVED** [review]
All of my suggestions are merely suggestions. I'm going to approve so we can keep moving forward quickly.

File: config-spec.md (hunk 3 above)
Comment (hktouw) [reply config-spec.md:109]: Perhaps? Maybe we can make this such that we give the function / component the config, cursor, and other relevant vars and it returns what frames to render? 

We'll have to play around with that though and see if that makes sense. I'm 👍 with going with your change for now and we can update the annotation to something else if that doesn't end up working.

File: config-spec.md (hunk 4 above)
Comment (hktouw) [reply config-spec.md:123]: Sounds good to me 👍

File: config-spec.md [hunk 5]
```diff
@@ -65,3 +67,76 @@ workflow:
...
+        pageSize: 50
+        source: mysql1
+      frames:
+        - frame:
+          view:
+            component: OTStoredQueryToolbar
+        - frame:
+          view:
+            component: TableOfContent
+              content_ids: :issue_ids # example of argument aliasing
```
Comment (hktouw) [inline config-spec.md:138]: @kmcq here is an example of your aliasing idea that we can do in processProps.

File: config-spec.md [hunk 6]
```diff
@@ -65,3 +67,76 @@ workflow:
...
+        props:
+          fields:
+            - :account_id
+    - frame:
+      data:
+        query: >\n
+          SELECT id AS issue_id
+          FROM issues
+          WHERE author_id = :account_id
+        params:
```
Comment (hktouw) [inline config-spec.md:123]: Per our conversation, we're doing to just declare parameters here as opposed to trying to extract them from the query. Furthermore, we don't have any type information in the query and couldn't do this anyway.

Pie in the sky idea 💡. In the distant future when we support integration with data catalogs, I wonder if we could get type information from that and apply it to the query? I feel like having queries dynamically typed would be a great feature. Another way we could do this too perhaps would be to run explain queries and extract type information that way.

**Comment by hktouw** [comment]
Going to merge this for now since this is a great start and we have something to work with. I'll open up issues for the various upgrades to the rendering engine that need to be made to support this config spec.

**Pull request merged**